src/Streamly/Internal/Data/Stream/StreamD.hs

```diff
@@ -3230,36 +3240,36 @@ tapRate ::
 tapRate samplingRate action (Stream step state) = Stream step' Nothing
   where
     {-# NOINLINE loop #-}
-    loop countRef prev = do
+    loop countVar prev = do
         i <-
             MC.catch
                 (do liftIO $ threadDelay (round $ samplingRate * 1000000)
-                    i <- liftIO $ readIORef countRef
+                    i <- liftIO $ readVar countVar
                     let !diff = i - prev
                     void $ action diff
                     return i)
                 (\(e :: AsyncException) -> do
-                     i <- liftIO $ readIORef countRef
+                     i <- liftIO $ readVar countVar
                      let !diff = i - prev
                      void $ action diff
                      throwM (MC.toException e))
-        loop countRef i
+        loop countVar i
 
     {-# INLINE_LATE step' #-}
     step' _ Nothing = do
-        countRef <- liftIO $ newIORef 0
-        tid <- fork $ loop countRef 0
+        countVar <- liftIO $ newVar 0
+        tid <- fork $ loop countVar 0
```

Comment:
Use `forkManaged` from the other PR?

Reply:
we can remove `tapRate` itself later, it can be expressed in terms of those new combinators.